### ISSUE TITLE
How to calculate integrals with vector delta functions

### ISSUE THREAD

**Issue by brandon-lab** [mfem/mfem#1287]
Hello, 

I am wondering how to calculate the following two integrals in mfem:

(1) `V = \int N \nabla \cdot J d\Omega`, where N is the H1 scalar basis function
(2) `V = \int J \cdot \nabla \times N d\Omega`, where N is the H(curl) vector basis function

J is a vector delta function with known location and direction, and J is local inside the element `\Omega`.

Thanks.

Brandon

**Comment by tzanio** [comment]
Just to clarify, the two linear forms you are interested in are: (J, ∇ u) and (J, ∇ x v), where J is a vector delta-function, and u and v are test functions in H1 and H(curl) respectively?

Currently there are no such integrators in MFEM, so you need to implement them yourself. We do have delta-function support in the classes derived from [DeltaLFIntegrator](https://github.com/mfem/mfem/blob/3032b74016c505046361caa372b4f2a0a5cba562/fem/lininteg.hpp#L47) (e.g. for (J,v)), so that could be a start.

Hopefully other developers or users can help more.

Tzanio

**Comment by psocratis** [comment]
Hi @brandon-lab. 

I have implemented (J, ∇ x v) in past and it's in [gmg-solver] branch: https://github.com/mfem/mfem/tree/gmg-solver in fem/lininteg.[ch]pp, and it is used in /examples/solvers-dev/FOSLS_maxwell.cpp for the case where J is a VectorFunctionCoefficient.
The case where J is a DeltaFunctionCoeffient  is there but I haven't tested it yet, so I would recommend to use it with caution. 

As @tzanio mentioned, there is no implementation of  (J, ∇ u), but It should be relatively straight forward to implement it, following the implementation of existing LF integrators. 

I hope this helps. 

Socratis

**Comment by brandon-lab** [comment]
Hi @psocratis ,

Thanks for providing your implementation. I tried the following codes:

```
/// \f$ (f, curl v)_{\Omega} \f$ for VectorFiniteElements (Nedelec, Raviart-Thomas)
class VectorFEDomainLFCurlIntegrator : public DeltaLFIntegrator
{
private:
   VectorCoefficient &QF;
   DenseMatrix vshape;
   Vector vec;

public:
   VectorFEDomainLFCurlIntegrator(VectorCoefficient &F) : DeltaLFIntegrator(F), QF(F) {}

   virtual void AssembleRHSElementVect(const FiniteElement &el, ElementTransformation &Tr, Vector &elvect);

   virtual void AssembleDeltaElementVect(const FiniteElement &fe, ElementTransformation &Trans, Vector &elvect);

   using LinearFormIntegrator::AssembleRHSElementVect;
};

// Class for domain integration of L(v) := (f, grad v), where
//    f=(f1,...,fn) and v=(v1,...,vn).
class VectorDomainLFGradIntegrator : public DeltaLFIntegrator
{
private:
   VectorCoefficient &QF;
   DenseMatrix dshape;
   Vector vec;

public:
   /// Constructs a domain integrator with a given VectorCoefficient
   VectorDomainLFGradIntegrator(VectorCoefficient &F) : DeltaLFIntegrator(F), QF(F) {}

   /** Given a particular Finite Element and a transformation (Tr)
        computes the element right hand side element vector, elvect. */
   virtual void AssembleRHSElementVect(const FiniteElement &el, ElementTransformation &Tr, Vector &elvect);

   virtual void AssembleDeltaElementVect(const FiniteElement &fe, ElementTransformation &Trans, Vector &elvect);

   using LinearFormIntegrator::AssembleRHSElementVect;
};

```

```
void VectorFEDomainLFCurlIntegrator::AssembleRHSElementVect(const FiniteElement &el, ElementTransformation &Tr, Vector &elvect)
{
   int dof = el.GetDof();
   int spaceDim = Tr.GetSpaceDim();

   vshape.SetSize(dof, spaceDim);
   vec.SetSize(spaceDim);

   elvect.SetSize(dof);
   elvect = 0.0;

   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      // int intorder = 2*el.GetOrder() - 1; // ok for O(h^{k+1}) conv. in L2
      int intorder = 2 * el.GetOrder();
      ir = &IntRules.Get(el.GetGeomType(), intorder);
   }

   for (int i = 0; i < ir->GetNPoints(); i++)
   {
      const IntegrationPoint &ip = ir->IntPoint(i);

      Tr.SetIntPoint(&ip);
      el.CalcPhysCurlShape(Tr, vshape);

      QF.Eval(vec, Tr, ip);
      vec *= ip.weight * Tr.Weight();

      vshape.AddMult(vec, elvect);
   }
}

void VectorFEDomainLFCurlIntegrator::AssembleDeltaElementVect(const FiniteElement &fe, ElementTransformation &Trans, Vector &elvect)
{
   MFEM_ASSERT(vec_delta != NULL, "coefficient must be VectorDeltaCoefficient");
   int dof = fe.GetDof();
   int spaceDim = Trans.GetSpaceDim();

   vshape.SetSize(dof, spaceDim);
   fe.CalcPhysCurlShape(Trans, vshape);

   vec_delta->EvalDelta(vec, Trans, Trans.GetIntPoint());

   elvect.SetSize(dof);
   vshape.Mult(vec, elvect);
}

void VectorDomainLFGradIntegrator::AssembleRHSElementVect(const FiniteElement &el, ElementTransformation &Tr, Vector &elvect)
{
   int dof = el.GetDof();
   int spaceDim = Tr.GetSpaceDim();

   dshape.SetSize(dof, spaceDim);

   elvect.SetSize(dof);
   elvect = 0.0;

   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      int intorder = 2 * el.GetOrder();
      ir = &IntRules.Get(el.GetGeomType(), intorder);
   }

   for (int i = 0; i < ir->GetNPoints(); i++)
   {
      const IntegrationPoint &ip = ir->IntPoint(i);

      Tr.SetIntPoint(&ip);
      el.CalcPhysDShape(Tr, dshape);

      QF.Eval(vec, Tr, ip);
      vec *= ip.weight * Tr.Weight();

      dshape.AddMult(vec, elvect);
   }
}

void VectorDomainLFGradIntegrator::AssembleDeltaElementVect(const FiniteElement &fe, ElementTransformation &Trans, Vector &elvect)
{
   MFEM_ASSERT(vec_delta != NULL, "coefficient must be VectorDeltaCoefficient");
   int dof = fe.GetDof();
   int spaceDim = Trans.GetSpaceDim();

   dshape.SetSize(dof, spaceDim);
   fe.CalcPhysDShape(Trans, dshape);

   vec_delta->EvalDelta(vec, Trans, Trans.GetIntPoint());

   elvect.SetSize(dof);
   dshape.Mult(vec, elvect);
}
```

The `class VectorFEDomainLFCurlIntegrator` is copied from your [gmg-solver] branch. Am I doing the right thing?

Best,
Brandon

**Comment by psocratis** [comment]
Hi @brandon-lab, yes i think it looks good. Can you test them and let us know how it goes? You can then open a PR and we can add them to the library

Thanks, 

Socratis

**Comment by brandon-lab** [comment]
I don't know if this would solve my problem. I am still debuging, currently I have not got the results as expected.

**Comment by brandon-lab** [comment]
Hi @psocratis , I calculate the fist integral by hand, and the result agrees with mfem. I think it is correct. Maybe I could open a PR soon.

**Comment by brandon-lab** [comment]
Hi,

I am now solving a problem involving two type of unknowns, p and q. The final system matrix is as follows:

```
\begin{equation}
{
\left[ \begin{array}{ccc}
A & B\\
C & D\\
\end{array} 
\right ]
}
{
\left[ \begin{array}{ccc}
p \\
q \\
\end{array} 
\right ]
}
=
{
\left[ \begin{array}{ccc}
m \\
n \\
\end{array} 
\right ]
}
\end{equation}
```
Both p and q satisfy a Dirichlet boundary, i.e., p=0, q=0 on the boundary. I assemble A and D by 
`pblf_a->FormSystemMatrix(ess_tdof_list_p, A);`
`pblf_d->FormSystemMatrix(ess_tdof_list_q, D);`
for B and C:
`pblf_b->FormRectangularSystemMatrix(ess_tdof_list_p,ess_tdof_list_q,B);`
`pblf_c->FormRectangularSystemMatrix(ess_tdof_list_q,ess_tdof_list_p,C);`

I have not changed the RHS m and n. Will this impose the boundary conditions for both p and q?

Brandon

**Comment by psocratis** [comment]
Hi @brandon-lab , 

The first two calls will eliminate (zero out) the columns and rows of the A and D corresponding to the entries listed in ess_tdof_list_p and ess_tdof_list_p respectively, and the diagonal entries will become either 1 or will not change depending on the diagonal policy. In the parallel case the diag entries will always be set to 1. 

The FormRectangularSystemMatrix calls will eliminate the entire row and column corresponding to the two lists provided. The solution vector or the right hand side are not modified with any of those calls. Instead you should use the FormLinearSystem/FormRectangularLinearSystem  method to impose the  BCs

Hope this helps,

Socratis

**Comment by jandrej** [comment]
As a little side note, be aware that FormRectangularLinear* methods currently are not working in parallell. This is fixed in #1259.
